### DISCUSSION TITLE
If I want to write a CLI application...

### DISCUSSION THREAD

**Discussion by atrauzzi** [microsoft/aspire#4625] [Q&A] [answered]
Is there any advice that I can follow so that the application can consume everything abstracted by Aspire?  Like say I have a database and I want to run some maintenance tasks.  Or maybe even use EFs migration bundling...

(etc) :slightly_smiling_face:

**Comment by davidfowl** [comment]
What do you mean exactly? That you want to write a cli application that uses the aspire components? Or something else?

**Reply by atrauzzi** [reply]
I think that's the same thing, yeah.  So right now I write CLI applications that can perform one-off tasks for me.

I can run them locally, or in prod environments, set these up using Cloud Run Jobs / Azure Container App Jobs to be triggerable.

Of course, if I'm going to run a CLI application on my workstation that's running things through aspire, perhaps I can leverage discovery and configuration?

**Reply by jack775544** [reply, marked as answer]
From what I have seen, Aspire components are built using the `IHostApplicationBuilder` which is not limited to web apps, but can also be used for console apps. You can read more about how to use it for a console app [here](https://learn.microsoft.com/en-us/dotnet/core/extensions/generic-host?tabs=appbuilder).

If you want a short lived appplication, you can do something like this
```cs
var builder = Host.CreateApplicationBuilder();

// This is stuff from Aspire
builder.AddServiceDefaults();
builder.AddNpgsqlDbContext<MyDbContext>("Postgres");

using var host = builder.Build();

var dbContext = host.Services.GetRequiredService<MyDbContext>();
// Do stuff here
```

Or if you want a long lived application, you can [implement a hosted service](https://learn.microsoft.com/en-us/dotnet/core/extensions/timer-service) and run the host instead of just building it.

**Reply by atrauzzi** [reply]
Works for me!  Thank you :slightly_smiling_face:
